apps/leo_storage/src/leo_storage_handler_del_directory.erl

```diff
@@ -0,0 +1,634 @@
+    _ = check_stats(self(), State),
+    {_, State_1} = dequeue(State),
+
+    %% Retrieve the latest items
+    State_2 = case leo_backend_db_api:fetch(
+                     ?DEL_DIR_STATE_DB_ID, <<>>,
+                     fun(K,_V, Acc) ->
+                             [K| Acc]
+                     end) of
+                  {ok, Dirs} ->
+                      State_1#state{cached_items = Dirs};
+                  not_found ->
+                      State_1#state{cached_items = []};
+                  {error, Cause} ->
+                      ?error("handle_info/2 - trigger", [{cause, Cause}]),
+                      State_1
+              end,
+    erlang:send_after(?INTERVAL, self(), trigger),
+    {noreply, State_2};
+
+handle_info({next, MQId, Directory, 0}, State) ->
+    %% Finished
+    case leo_backend_db_api:get(?DEL_DIR_STATE_DB_ID, Directory) of
+        {ok, DelDirStateBin} ->
+            case catch binary_to_term(DelDirStateBin) of
+                #del_dir_state{mq_id = MQId,
+                               type = Type} ->
+                    run(Type, ?STATE_FINISHED, MQId, Directory, State);
+                {'EXIT', Cause} ->
+                    ?error("handle_info/2 - next", [{cause, Cause}])
+            end;
+        not_found ->
+            ok;
+        {error, Cause} ->
+            ?error("handle_info/2 - next", [{cause, Cause}])
+    end,
+    {noreply, State};
+
+handle_info({next,_MQId,_Directory,_TotalMsgs}, State) ->
+    {noreply, State};
+
+handle_info({failed, MQId, Type, Directory}, State) ->
+    update_state(#del_dir_state{
+                    mq_id = MQId,
+                    directory = Directory,
+                    type = Type,
+                    state = ?STATE_PENDING,
+                    timestamp = leo_date:now()}),
+    {noreply, State};
+
+handle_info({ongoing, MQId, Type, Directory}, State) ->
+    {ok, State_1} = run(Type, ?STATE_ONGOING, MQId, Directory, State),
+    {noreply, State_1};
+
+handle_info({enqueued, MQId, Type, Directory}, State) ->
+    {ok, State_1} = run(Type, ?STATE_MONITORING, MQId, Directory, State),
+    {noreply, State_1};
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+
+%% @doc This function is called by a gen_server when it is about to
+%%      terminate. It should be the opposite of Module:init/1 and do any necessary
+%%      cleaning up. When it returns, the gen_server terminates with Reason.
+terminate(_Reason,_State) ->
+    ok.
+
+
+%% @doc Convert process state when code is changed
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+
+%%====================================================================
+%% Private Functions
+%%====================================================================
+%% @private
+-spec(dequeue(State) ->
+             {ok, State} when State::#state{}).
+dequeue(#state{num_of_workers = NumOfWorkers} = State) ->
+    case get_ongoing_workers(NumOfWorkers) of
+        {ok, DelBucketStateList} ->
```

Comment:
Maybe commenting about what the below listing does (retrieving available mq_id) would be good for those who try to read the code for the first time.